src/cmds/qstat.c

```diff
@@ -1609,6 +1611,41 @@ display_statjob(struct batch_status *status, struct batch_status *prtheader, int
 }
 
 #define TYPEL 4
+/**
+ * @brief
+ *      Helper function to accumulate new-type restriction to a running string
+ *
+ * @param[in] keys - array of  keys
+ * @param[in] values - array of values corresponding to each key
+ * @param[in] count - pointer to current number of unqiue keys stored in keys/values
+ * @param[in] max - maximum allowed number of unique keys 
+ * @param[in] key - new key to process 
+ * @param[in] val - value associated with the key to accumulate 
+ *
+ */
+static void accumulate_restriction(char **keys, char **values, int *count, int max, const char *key, const char *val){
+	for (int i = 0; i < *count; i++){
+		// check if key already exists
+		if(strcmp(keys[i], key) == 0){
+			size_t new_len = strlen(values[i]) + strlen(val) + 2; // ",\0"
+			char *new_buf = malloc(new_len);
+			if (!new_buf){
+				exit_qstat("out of memory");
+			}
+			snprintf(new_buf, new_len, "%s,%s", values[i], val); //append to running output
+			free(values[i]);
+			values[i] = new_buf;
+			return;
+		}
+	}
+
+	// first time restriction seen, create new key
+	if (*count < max){
+		keys[*count] = strndup(key, strlen(key));
+		values[*count] = strndup(val, strlen(val));
```

Comment:
Shouldn't you be checking if the strndup() fail and bail out? I see the *count is incremented which is bad because another function would look into 'keys' and 'values' and could be processing NULL entries.